src/App.tsx

```diff
@@ -1,26 +1,199 @@
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const inputRef = useRef<HTMLInputElement | null>(null);
+  const [deletingTodos, setDeletingTodos] = useState<number[]>([]);
+  const [errorMessage, setErrorMessage] = useState<string>('');
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(data => {
+        setTodoList(data);
+        setFilteredTodoList(data);
+      })
+      .catch(() => {
+        setErrorMessage('Unable to load todos');
+        setTimeout(() => setErrorMessage(''), 3000);
+      });
+  }, []);
+
+  useEffect(() => {
+    if (inputRef.current) {
+      inputRef.current.focus();
+    }
+  }, [tempTodo, deletingTodos]);
+
+  const filterTodos = useCallback(
+    (status: string): void => {
+      setSelectedFilter(status);
+
+      switch (status) {
+        case 'active':
+          setFilteredTodoList(todoList.filter(todo => !todo.completed));
+          break;
+        case 'completed':
+          setFilteredTodoList(todoList.filter(todo => todo.completed));
+          break;
+        case 'all':
+        default:
+          setFilteredTodoList(todoList);
+          break;
+      }
+    },
+    [todoList],
+  );
+
+  useEffect(() => {
+    filterTodos(selectedFilter);
+  }, [todoList, filterTodos, selectedFilter]);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  function handleToggleCompletion(todoId: number) {
+    setTodoList(currentTodos =>
+      currentTodos.map(todo =>
+        todo.id === todoId ? { ...todo, completed: !todo.completed } : todo,
+      ),
+    );
+  }
+
+  function handleChangeInput(e: React.ChangeEvent<HTMLInputElement>) {
+    setTitle(e.target.value);
+  }
+
+  function submitTodo(e: React.FormEvent) {
+    e.preventDefault();
+
+    if (title.trim() === '') {
+      sendErrorMessage('Title should not be empty', setErrorMessage);
+
+      return;
+    }
+
+    const newTempTodo: Todo = {
+      id: 0,
+      userId: USER_ID,
+      title: title.trim(),
+      completed: false,
+    };
+
+    setTempTodo(newTempTodo);
+    setLoader(true);
+
+    postTodo({
+      userId: USER_ID,
+      title: title.trim(),
+      completed: false,
+    })
+      .then(data => {
+        setTodoList(preTodoList => [...preTodoList, data]);
+        setTempTodo(null);
+        setTitle('');
+      })
+      .catch(() => {
+        setTempTodo(null);
+        sendErrorMessage('Unable to add a todo', setErrorMessage);
+      })
+      .finally(() => {
+        setLoader(false);
+      });
+  }
+
+  async function handleDeleteTodo(todoId: number) {
+    setDeletingTodos(prev => [...prev, todoId]);
+
+    try {
+      await deleteTodo(todoId);
+      setTodoList(prevTodoList =>
+        prevTodoList.filter(todo => todo.id !== todoId),
+      );
+    } catch {
+      sendErrorMessage('Unable to delete a todo', setErrorMessage);
+    } finally {
+      setDeletingTodos(prev => prev.filter(id => id !== todoId));
+    }
+  }
+
+  async function handleDeleteCompletedTodo() {
```

Comment:
It is important to follow the same way of creating functions everywhere
```suggestion
  const handleDeleteCompletedTodo = async () => {
```